### ISSUE TITLE
Property types support parameter validation/rejection at configuration time

### ISSUE THREAD

**Issue by mkobit** [gradle/gradle#4468]
**Note**: I don't know if this can be just be supported in Groovy/Kotlin setter overriding or overriding, but wanted to bring it up here first. I also don't know what the idiomatic Gradle way of handling property validation is.

Using the new `Property` types, I would like to provide feedback early to users by creating a `Property` with validation for the boxed type. (similar to Kotlin's https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.properties/-delegates/vetoable.html ). One possible issue here is the [`set(Provider<? extends T> provider)`](https://docs.gradle.org/current/javadoc/org/gradle/api/provider/Property.html#set-org.gradle.api.provider.Provider-) method which may muddy this request a bit more. Other options is to perform during execution or side-effecting `map {}` step,  which are also fine. I just wanted to bring this case up as a UX issue for consumers of `Property` based plugins.

### Expected Behavior

As a plugin author, I would like to perform validation during configuration to give faster feedback to users.

### Current Behavior

Validation needs to be deferred to when the value is extracted.

### Context

A user trying to configure a `Property<String>` did not know the right "pattern" to configure it. It took several tries for the user to get right because of how it was internally being mapped into downstream `Provider<>` but was not getting validated at the source.

### Your Environment

Gradle 4.5.1

**Comment by ajoberstar** [comment]
Yeah, this was one of the nice features of the now-deprecated model API. A lifecycle around the model objects that let you put validation in before the values got used.

**Comment by mkobit** [comment]
@ajoberstar thanks, that is the closest to what I was trying to describe - the "at configuration time" part isn't  important to me, but providing a solid user experience for various `Property<>`/`Provider<>` type inputs in tasks/extensions/domain objects/wherever else is.

**Comment by adammurdoch** [comment]
This is a good use case. I think we should certainly add some way for a plugin to attach rules to validate property values.

**Comment by lptr** [comment]
This could be something we could incorporate in our own validation for task (and other) properties.

**Comment by bmuskalla** [comment]
We have a similar usecase [here](https://github.com/gradle/gradle/pull/11568) about using a `Property<Integer>` that should only accept positive values.

**Comment by ov7a** [comment]
This feature request is in the backlog of the relevant team and is prioritized by them.

Related:
* https://github.com/gradle/gradle/issues/29256
